feat(core): ship some types

boneskull opened the pull request:
This adds _some_ types to `lavamoat-core`.

Interesting stuff includes the `LavaMoatPolicy` type and the types for `endowmentsToolkit.js`

legobeat commented on packages/core/package.json:
```diff
@@ -12,9 +12,11 @@
     "lavamoat-tofu": "^6.2.1",
-    "merge-deep": "^3.0.3"
+    "merge-deep": "^3.0.3",
+    "type-fest": "^4.3.3"
```
Can we achieve this without needing a new runtime dependency? Seems like this belongs in devDependencies.

boneskull replied:
It's not a dev dep because we export a type that uses it (policy schema).

legobeat commented on packages/core/src/endowmentsToolkit.js:
```diff
@@ -328,12 +418,19 @@ function endowmentsToolkit({
     while (true) {
       // abort if this is the end of the prototype chain.
-      if (!receiver) {
+      if (receiver === null) {
```
Potential behavior change, I think? Can those be broken out as separate PR(s) and merged independently to shipping types?

boneskull replied:
I've reverted this

naugtur commented on packages/core/src/endowmentsToolkit.js:
```diff
@@ -1,12 +1,26 @@
-// the contents of this file will be copied into the prelude template
-// this module has been written so that it required directly or copied and added to the template with a small wrapper
```
We should keep this comment, unless it's no longer true.

boneskull replied:
it's there; it's just moved.

naugtur commented on packages/core/src/generatePolicy.js:
```diff
@@ -126,7 +170,7 @@ function createModuleInspector(opts = {}) {
         errorRecovery: true,
       })
-    if (includeDebugInfo && ast.errors && ast.errors.length) {
+    if (includeDebugInfo && isParsedAST(ast) && ast.errors.length) {
```
This is clever. And I don't even mind the runtime overhead vs clarity in this case.

boneskull replied:
`ast.errors` doesn't mean very much to TS' type inference, but a proper type guard like `isParsedAST()` does.

naugtur commented on packages/core/src/generatePolicy.js:
```diff
@@ -246,16 +316,28 @@ function createModuleInspector(opts = {}) {
+   * @type {GeneratePolicyFn}
+   */
+  function generatePolicy({ policyOverride, includeDebugInfo = false }) {
+    /** @type {import('./schema').Resources} */
```
we're in a closure so if I understood correctly, you can do a typedef import and get ResourcePolicy etc. directly without reexporting it. Was that the case that only top level typedef needs to be avoided because it's reexported?

boneskull replied:
Yes, it could be used here without worrying about an export (I think!), but if `Resources` becomes generic in the future, we would need to re-implement those  type args in `@template` tags.  That said, you could probably get away with it here--kind of like you can get away with writing JS without semicolons.

_Personally,_ I don't want to think too hard about whether or not it's safe to create the `@typedef` reference, so I choose not to and make that tradeoff.  Similarly, I don't like thinking about whether or not ASI is going to work either, so I always use semicolons. :smile:

naugtur replied:
As the best practices around this are far from set in stone, I'm trying to explore options in case we come up with something that's more convenient. 

Also note that making this pattern more popular makes SES slower, because it relies on a transform to replace each `import(` with something else

boneskull replied:
The transform in question doesn't happen in SES, fwiw.  It means we have to apply it ourselves; hence https://github.com/endojs/endo/pull/1812

kumavis commented on packages/core/src/endowmentsToolkit.js:
```diff
@@ -379,7 +484,7 @@ function endowmentsToolkit({
       const descriptors = Object.getOwnPropertyDescriptors(source)
       Object.values(descriptors).forEach((desc) => {
-        if ('get' in desc) {
+        if ('get' in desc && desc.get) {
```
I usually do `desc.get !== undefined`

boneskull replied:
i want it to be truthy

boneskull replied:
fwiw `'foo' in bar` is the TypeScript-Approved:tm: way to check if a prop "has" something; skipping right to `foo.bar` will typically be an error if it's unclear from type inference

naugtur replied:
in the case of a descriptor there's no difference between it being truthy or not-undefined because it can only be a function or missing. 
Is this for typescript only?

boneskull replied:
According to TS, `get` can be `null`.  And `get: null` would be true for `'get' in desc`; hence the truthiness check.  TS' ECMAScript libs aren't the spec, but I trust them.

naugtur replied:
If we want to support the folks using typescript directly on platforms we don't yet foresee (poor souls) this seems like the right thing to do.

kumavis commented on packages/core/src/generatePolicy.js:
```diff
@@ -160,6 +213,7 @@ function createModuleInspector(opts = {}) {
     if (includeDebugInfo) {
       const moduleDebug = debugInfo[moduleRecord.specifier]
       moduleDebug.sesCompat = {
+        // XXX: I don't think this is needed, since it appears we overwrite all properties
```
this is just for the debug file. what do you mean about overwriting all props?

boneskull replied:
`compatWarnings` has props `primordialMutations`, `strictModeViolations` and `dynamicRequires`, all of which are overwritten below.  I don't think there are other props.

boneskull replied:
(meaning that `...compatWarnings` is unnecessary. at least, I think)

boneskull committed:
feat(core): ship some types

This adds some (not all!) types to `lavamoat-core` and generates type declarations.

## Regarding Types

- Only sources with the `// @ts-check` directive are type-checked by `tsc` when building declarations.
- Types added should be considered "best effort" given my assumptions of intent.
- Any source changes which change behavior (there are few) should be considered opportunistic bug fixes.
- TypeScript does not like variable reassignment; if we reassign to a variable and the new type is different from the initial type, it will complain, and I had to avoid this by creating new variables
- The LavaMoat policy schema has a versioning scheme, only because we may want to introduce policy file versioning in the future.
- `src/schema/ambient.ts` is an abberation, and its reason for existence can only be avoided by applying AST-based transforms against our sources.  Any code which gets pulled into a SES compartment will need to consume global types and won't be able to use `import()` in docstrings.  Unfortunately, it's not immediately obvious which files do and don't end up in a compartment.

## Regarding dependencies or `package.json`

- `type-fest` is a prod dependency of `lavamoat-core` because one or more modules exports an API referencing `type-fest`'s types.  `@types/json-stable-stringify` is used, but there is no module exporting an API depending on types from that package.
- `src/types.ts` only exists to emit `types/src/types.d.ts`, which is referenced as the package entry point for TS (see `types` in `package.json`).
- A `files` prop was added to the `package.json` as an "allow list" of files to publish.  `tsc` leaves `*.tsbuildinfo` files hanging around for its incremental builds, and we don't need to publish them.

## Other

- Two slight tweaks in our ESLint config were necessary to make it pass.  Linting of `lavamoat-core` is not generally leveraging `@typescript-eslint`--which provides more coverage--which will happen in another PR.

naugtur commented on packages/core/src/endowmentsToolkit.js:
```diff
@@ -263,20 +329,29 @@ function endowmentsToolkit({
           receiver === unwrapFromGlobalThis ? unwrapToGlobalThis : receiver
         // sometimes getters replace themselves with static properties, as seen wih the FireFox runtime
-        const result = Reflect.apply(sourcePropDesc.get, receiverRef, [])
+        const result = Reflect.apply(getter, receiverRef, [])
```
(must) Let's make sure this is not breaking the code. I'm pretty sure the comment above is explaining why we're not using a local copy but referencing the getter from the original source and this change would break it.

legobeat replied:
Regardless of outcome:  This seems like a non-type-related refactor that could/should be broken out as separate change? Or maybe I'm missing some context on why this change is here?

boneskull replied:
Sorry this wasn't explained; TS' type inference doesn't know that `sourcePropDesc.get` is not `undefined` at this point, because it doesn't know when `wrappedPropDesc.get()` is called, so it cannot be sure, despite the conditional on L345 (or L346, depending).

I've changed it to just use a type assertion instead.

To be clear, I didn't make any code changes in this PR that _weren't_ prompted by type checking.

boneskull replied:
By leaving it as-is, we're saying: _`sourcePropDesc.get` is allowed to be `undefined` here, and if so, allow `Reflect.apply()` to throw an exception_.  Just want to confirm that's the intent.

naugtur replied:
We could doublecheck before applying. The breaking change was that we cached the getter instead of using the fresh one that could have been updated.

boneskull replied:
I'm not sure what sort of action would be reasonable here _other_ than just letting it throw.  In other words, I don't know how to improve on that.

naugtur reviewed (changes requested):
I've spotted a new bug being introduced, I think: 
https://github.com/LavaMoat/LavaMoat/pull/732/files#diff-5ed0581fdb5b3ec8105f7f2e6773d6f5db45dbe74dccf2a921cc9c0e40ecda60R354

naugtur commented on packages/core/src/endowmentsToolkit.js:
```diff
@@ -295,17 +379,26 @@ function endowmentsToolkit({
         const receiverRef =
           receiver === unwrapFromGlobalThis ? unwrapToGlobalThis : receiver
-        return Reflect.apply(sourcePropDesc.set, receiverRef, [value])
+        return Reflect.apply(setter, receiverRef, [value])
```
(must) same issue as with https://github.com/LavaMoat/LavaMoat/pull/732/files#diff-5ed0581fdb5b3ec8105f7f2e6773d6f5db45dbe74dccf2a921cc9c0e40ecda60R354

Sorry I didn't point it out here. I did in a different PR that was showing it. 

GH will not let me add a suggestion that fixes it...

boneskull replied:
done

boneskull committed:
chore(core): peer review fix
